Slider component for @react-stately and @react-aria

chungwu opened the pull request:
This creates:

* `useMultiSliderState` for @react-stately/slider for managing slider state for multiple values
* `useMultiSlider` for @react-aria/slider for managing interactions and accessibility for slider components

Includes some Storybook stories and toy components using the above hooks to build sliders.

Does _not_ include slider implementations for @react-aria/spectrum.

Closes #763

## ✅ Pull Request Checklist:

- [X] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [X] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [X] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

For now, just visual tests on Storybook under the "Slider" group, starting here: http://localhost:9003/?path=/story/slider--single

## 🧢 Your Project:

[Plasmic](https://plasmic.app)

devongovett commented:
Skimmed this just now and it looks awesome! Will take a more in depth look tomorrow. 😍

devongovett commented on packages/@react-aria/slider/src/useMultiSlider.ts:
```diff
@@ -0,0 +1,105 @@
+            e.preventDefault();
+
+            realTimeTrackDraggingIndex.current = index;
+            state.setFocusedIndex(index);
```
Do you think we should also set the value of the slider here as well? Currently, if you mouse down on the track without moving your mouse, the thumb does not move. It only moves after you drag the mouse.

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,124 @@
+  state: MultiSliderState,
+  trackRef: React.RefObject<HTMLElement>,
+  inputRef: React.RefObject<HTMLInputElement>
+): SliderThumbAria {
```
This is starting to be a lot of arguments. What do you think about making an options object instead, and include the necessary props from slider props and thumb props, along with the two refs. The state could continue being a separate argument potentially.

```tsx
interface SliderThumbOptions extends SliderThumbProps {
  maxValue?: number,
  minValue?: number,
  step?: number,
  isReadOnly?: boolean,
  isDisabled?: boolean,
  trackRef: React.RefObject<HTMLElement>,
  inputRef: React.RefObject<HTMLInputElement>
}

useSliderThumb(options: SliderThumbOptions, state: SliderState)
```

chungwu replied:
Ah, agreed making `SliderThumbOptions` is nicer.  However, I'm leaning towards keeping thumb and slider props as separate fields, rather than having it extend `SliderThumbProps`, to be extra explicit:

```typescript
interface SliderThumbOptions {
  sliderProps: BaseSliderProps;
  thumbProps: SliderThumbProps;
  trackRef: React.RefObject<HTMLElement>;
  inputRef: React.RefObject<HTMLInputElement>;
}
```

devongovett replied:
I think we'd prefer to combine the two props objects. I had a suggestion to avoid needing `minValue` and `maxValue` below, and the others could potentially be different per thumb rather than just a global setting. e.g. `isDisabled` and `isReadOnly` are available on individual thumbs, and perhaps `step` may as well eventually.

chungwu replied:
`isReadOnly` and `isDisabled` are actually already part of `SliderThumbProps`...  If we combine, then either we'd need to name the slider-level readonly/disabled differently:

```tsx
interface SliderThumbOptions extends SliderThumbProps {
  isSliderReadOnly?: boolean;
  isSliderDisabled?: boolean;
  step?: number;
  // etc
}
```

Or whoever is calling `useSliderThumb()` would need to take care to merge them:

```tsx
const {...} = useSliderThumb({
    ...thumbProps,
   isReadOnly: thumbProps.isReadOnly || sliderProps.isReadOnly,
   isDisabled: thumbProps.isDisabled || sliderProps.isDisabled,
   step: sliderProps.step,
});
```

This is actually why I originally leaned towards keeping `sliderProps` and `thumbProps` as separate fields in `SliderThumbOptions`.  But happy to go with your preference for either of the above here!

chungwu replied:
Also, I believe you'd still need minValue / maxValue, so that `useSliderThumb` can map from offset position to actual value (which it currently does by calling `computeOffsetToValue` with the sliderProps, though it really just needs min/max/step).

Another possibility is to get the sliderProps from `state` instead of having it passed in explicitly in `SliderThumbOptions`, since `SliderState` already knows about the slider props...

devongovett replied:
> Or whoever is calling useSliderThumb() would need to take care to merge them:

Yep, I think this is what I'd expect. It should be up to whatever is rendering the thumb to decide which thumbs are disabled IMO. We do this in e.g. Radio already, to combine the group prop with the individual prop.

> Another possibility is to get the sliderProps from state

Yeah, we do this in a few places already. Perhaps min/max/step could be stored as attributes on the state so they can be accessed by the thumbs? Or perhaps the state should have a method for converting from a percentage offset to a value between min and max. Maybe `state.setThumbPercent(index, percent)` - would correspond nicely with `getThumbPercent`.

chungwu replied:
Great idea!

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,124 @@
+      'aria-required': isRequired || undefined,
+      'aria-invalid': validationState === 'invalid' || undefined,
+      'aria-errormessage': thumbProps['aria-errormessage'],
+      onKeyDown: allowDrag ? draggableProps.onKeyDown : undefined
```
I think you might need an `onChange` event handler here. Right now, adjusting the value with a screen reader isn't working. For example, on iOS you can use swipe gestures with VoiceOver to increment or decrement the value. This would fire change events on the hidden input.

You might not need the keyboard handler in that case - the native input might already handle the keyboard events and fire the onChange event accordingly. We'd need to test cross browser to ensure it's consistent.

chungwu replied:
Great! Now the browser can also handle page-up/page-down as it sees fit

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,124 @@
+      'aria-errormessage': thumbProps['aria-errormessage'],
+      onKeyDown: allowDrag ? draggableProps.onKeyDown : undefined
+    }),
+    thumbProps: {
```
Maybe do something like

```tsx
thumbProps: allowDrag ? mergeProps(draggableProps, {
  onMouseDown: focusInput
}) : {}
```

That way any props that the drag hook returns will get passed through, not just these three.

devongovett commented on packages/@react-aria/slider/src/useMultiSlider.ts:
```diff
@@ -0,0 +1,105 @@
+ * accepts click and drag motions, so that the closest thumb will follow clicks and drags on
+ * the track.
+ */
+export function useMultiSlider(
```
I'm thinking we should just call this `useSlider`. I don't think we need to reserve that name for anything else, so many that's simpler?

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,124 @@
+      'aria-valuemax': state.getMaxValueForIndex(index),
+      'aria-readonly': isReadOnly || undefined,
+      'aria-disabled': isDisabled || undefined,
+      'aria-orientation': 'horizontal',
```
I guess this should come from a prop eventually

chungwu replied:
Yup; just starting with horizontal for now; should be straightforward to extend to vertical as well.

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,124 @@
+      tabIndex: allowDrag ? 0 : undefined,
+      'aria-valuenow': value,
+      'aria-valuemin': state.getMinValueForIndex(index),
+      'aria-valuemax': state.getMaxValueForIndex(index),
```
You could probably also set the `min`, `max` and `step` properties of the input as well. Perhaps these could represent the full range of the slider, while the aria values would represent the min/max for the current handle as you have it.

chungwu replied:
Hmm, I didn't find any best practices on multi-thumb sliders with range inputs re: min/max setting "full range" vs aria-valuemin/max...  It seems to make most sense to me for them to be the same (or, to prefer setting min/max over aria-valuemin/max), but happy to follow the guidance from you folks!

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,124 @@
+      'aria-valuemin': state.getMinValueForIndex(index),
+      'aria-valuemax': state.getMaxValueForIndex(index),
+      'aria-readonly': isReadOnly || undefined,
+      'aria-disabled': isDisabled || undefined,
```
probably also (instead?) set the native `disabled`/`readOnly` props to ensure the input isn't focusable if it's disabled.

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,124 @@
+    orientation: 'horizontal',
+    onDrag: (dragging) => {
+      state.setDragging(index, dragging);
+      console.log('THUMB FOCUS');
```
console

devongovett commented on packages/@react-types/slider/README.md:
```diff
@@ -0,0 +1,3 @@
+# @react-types/radio
```
```suggestion
# @react-types/slider
```

devongovett commented on packages/@react-stately/slider/src/useMultiSliderState.ts:
```diff
@@ -0,0 +1,116 @@
+    const newValues = replaceIndex(values, index, value);
+    setValues(newValues);
+
+    if (props.onChangeEnd && !realTimeDragging.current) {
```
Is `realTimeDragging` the same as `!isDraggings.some(Boolean)`? At the moment it's not, because if you were dragging multiple handles at the same time (e.g. multitouch), and stopped dragging one it would be set to false (below), even though you're still dragging one of the other handles. Perhaps these could be combined?

devongovett commented:
A couple other high level things:

1. Should handles be able to overlap? Noticed that you can drag them directly on top of one another. Should we limit dragging so it can only come within some minimum distance (or touching)? 
    <img width="87" alt="image" src="https://user-images.githubusercontent.com/19409/88252501-46cead80-cc63-11ea-8d16-789dac542654.png">

2. We'll need to adjust things so it works on touch devices. At the moment the handles don't move. I imagine this has more to do with the existing `useDrag1D` hook than anything you've implemented, so totally not your fault. And we can get this PR in as an initial implementation and follow up with touch support later as well. I think eventually we'll want to move that hook into `@react-aria/interactions` and add the touch support there. This will also apply to the mouse events that you've added in slider eventually as well.

Thanks again for this awesome (and quick) work! 😍

chungwu commented:
@devongovett Many thanks for the quick response!  Makes sense to me; will get back to this soon :sweat_smile: 

Re: overlapping handles -- this seems like an opinionated decision that perhaps should be made by the component author (so, for example, would live in react-spectrum and not react-aria)?  react-aria will give you the offset percentages of each handle, and it is up to you to decide how you want to render those handles...

devongovett commented:
That makes sense to me!

chungwu commented on packages/@react-aria/slider/src/useMultiSlider.ts:
```diff
@@ -0,0 +1,105 @@
+  trackProps: HTMLAttributes<HTMLElement>;
+
+  /** Label ID for labling slider inputs. Should be passed on in `useSliderThumb`. */
+  labelId: string;
```
Makes sense; I'll put this into a `commonThumbProps` object.

chungwu commented:
I've taken a pass at addressing the feedback.  Once there's consensus on the API and the behavior, I can start writing up some tests :smile:

devongovett commented on packages/@react-aria/slider/src/useSliderThumb.ts:
```diff
@@ -0,0 +1,137 @@
+    onIncrement: () => state.setValue(index, value + step),
+    onDecrement: () => state.setValue(index, value - step),
+    onIncrementToMax: () => state.setValue(index, maxValue),
+    onDecrementToMin: () => state.setValue(index, minValue)
```
I don't think these are working right now? Home and End aren't doing anything for me. I think you may need to pass through the `onKeyDown` handler to the input element?

devongovett replied:
Maybe we could add methods for incrementing/decrementing a thumb to stately? Then you wouldn't need to pass through `minValue` and `maxValue` as well. These are somewhat confusing since they are the minimum/maximum of the whole slider rather than the individual thumb. Setting the value only works because it is clamped in `state.setValue`. So maybe having a method like `state.incrementToMax(index)` would be clearer?

chungwu replied:
Hmm, weird; Home and End do work for me (Mac/Chrome, Mac/Safari, Linux/Chrome).  The focus is on the `input[type=range]` control, so as long as the browser handles those shortcuts, we _should_ see the changes in the onChange handler on the input... :thinking: 

We actually don't need to pass in onIncrement/onIncrementMax/etc. at all now, since we're just letting the browser handle the stepping...  Will remove!

chungwu replied:
(Could you let me know the browser/OS combo where you're seeing the Home/End issue?)

devongovett replied:
It was macOS safari. But let me check again.

devongovett commented on packages/@react-stately/slider/src/useSliderState.ts:
```diff
@@ -0,0 +1,117 @@
+
+  // Returns the value offset as a percentage from 0 to 1.
+  getOffsetPercentForIndex: (index: number) => number,
+  getOffsetPercentForValue: (value: number) => number,
```
These methods have potentially confusing names. Maybe combine them into a single one like `getThumbPercent(index: number, origin: number = minValue) => number`? Then you'd pass in an index of the thumb, and an optional origin to get the percentage from. What do you think?

chungwu replied:
I did want something more general that allows me to map any value to a percent though (not just the origin, which is just one example of this use case).  For example, maybe I'd like to render tick marks for different values on the track, and would need to know where to place them, etc.

That said, we don't have to hang this off of `SliderState` (though it is convenient).  We could instead have some top-level function like this:

```tsx
// maybe part of useSlider.ts?
export function getOffsetPercentForValue(props: BaseSliderProps, value: number): number {

}
```

There's also `SliderState.getValueLabelForValue()`, which, for the same reason, would be nice to have for rendering tick marks on the track.  This will have to be off of `SliderState` instead of a top-level function, though, since it relies on the `useNumberFormatter` hook.  I think if this one is on the `SliderState`, then probably makes sense to keep the family of similar functions on `SliderState` too.

Definitely open to better names though :sweat_smile:  Maybe we could do, `getThumbPercent(index: number)` vs `getValuePercent(value: number)`?

devongovett replied:
Makes sense. Those names sound good to me. 👍 

> There's also SliderState.getValueLabelForValue

 Maybe `getFormattedValue(value: number)`?

devongovett commented on packages/@react-stately/slider/src/useSliderState.ts:
```diff
@@ -0,0 +1,117 @@
+
+  // Returns the string label for the value, per props.formatOptions
+  getValueLabelForIndex: (index: number) => string,
+  getValueLabelForValue: (value: number) => string,
```
Maybe `getThumbValueLabel`? I think `getValueLabelForValue` could also be removed, looks like it's not used outside here?

devongovett commented on packages/@react-stately/slider/src/useSliderState.ts:
```diff
@@ -0,0 +1,117 @@
+
+  // Returns the min and max values for the index
+  getMinValueForIndex: (index: number) => number,
+  getMaxValueForIndex: (index: number) => number
```
`getThumbMinValue`/`getThumbMaxValue`?

chungwu commented on packages/@react-aria/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,117 @@
+
+  /** Partial props that should be passed into `useSliderThumb` for all thumbs.  
+   * Includes the Label ID for labeling slider inputs. */
+  commonThumbProps: CommonSliderThumbProps
```
Sounds good.  I called it `commonThumbProps` because it otherwise often interfered with unpacking the result from `useSliderThumb`, like so:

```tsx
const {thumbProps: commonThumbProps} = useSlider(...);
const {thumbProps} = useSliderThumb({ thumbProps: {...commonThumbProps, ...}});
```

But happy to adhere to convention!

chungwu commented:
Hi all!  Just took another pass addressing the feedback, as well as adding tests for `useSliderState`, `useSlider`, and `useSliderThumb`.  

The only piece of feedback I'm still hesitating about is this: https://github.com/adobe/react-spectrum/pull/809#discussion_r459204640  Again, happy to do what you believe is most conventional here, just wanted to point out a few things!

chungwu commented:
I believe I've resolved all the remaining comments (except for keyboard shortcuts not working for @devongovett when he tried it last).  Thanks for the feedback!

devongovett commented:
Excellent! Will take one final look tomorrow morning and get it merged. 😄

chungwu commented:
Yay!  It's clear you folks really prioritize community involvement, and working on this PR was an awesome experience :heart_eyes:
